### ISSUE TITLE
[REVIEW]: exoplanet: Gradient-based probabilistic inference forexoplanet data & other astronomical time series

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3285]
**Submitting author:** @dfm (<a href="http://orcid.org/0000-0002-9328-5652">Daniel Foreman-Mackey</a>)
**Repository:** <a href="https://github.com/exoplanet-dev/exoplanet" target ="_blank">https://github.com/exoplanet-dev/exoplanet</a>
**Version:** v0.5.1
**Editor:** @arfon
**Reviewer:** @grburgess, @benjaminpope
**Archive:** <a href="https://doi.org/10.5281/zenodo.5006965" target="_blank">10.5281/zenodo.5006965</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/978b4988d1c25768188b3642476eff2c/status.svg)](https://joss.theoj.org/papers/978b4988d1c25768188b3642476eff2c)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/978b4988d1c25768188b3642476eff2c"><img src="https://joss.theoj.org/papers/978b4988d1c25768188b3642476eff2c/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/978b4988d1c25768188b3642476eff2c/status.svg)](https://joss.theoj.org/papers/978b4988d1c25768188b3642476eff2c)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/exoplanet-dev/exoplanet" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@grburgess & @benjaminpope, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @arfon know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/exoplanet-dev/exoplanet">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@dfm) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/exoplanet-dev/exoplanet">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@dfm) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @grburgess, @benjaminpope it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by arfon** [comment]
@grburgess, @benjaminpope – This is the review thread for the paper. All of our communications will happen here from now on.

Please read the "Reviewer instructions & questions" in the first comment above.

Both reviewers have checklists at the top of this thread (in that first comment) with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention `https://github.com/openjournals/joss-reviews/issues/3285` so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for the review process to be completed within about 4-6 weeks but please make a start well ahead of this as JOSS reviews are by their nature iterative and any early feedback you may be able to provide to the author will be very helpful in meeting this schedule.

**Comment by grburgess** [comment]
@arfon I have had the same issue on other reviews... I am not able to check the boxes. It seems independent of the browser. Is this a common issue?

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.57 s (116.2 files/s, 15678.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          39           1170           1566           3968
TeX                              1             43              0            646
Markdown                         6             68              0            372
reStructuredText                 6            190             93            301
YAML                             9             33              0            272
CSS                              1             16              0             78
make                             1              8              5             30
TOML                             1              2              0             28
INI                              1              0              0              8
HTML                             1              0              0              6
-------------------------------------------------------------------------------
SUM:                            66           1530           1664           5709
-------------------------------------------------------------------------------


Statistical information for the repository '58c76c3075310f0d22a5303f' was
gathered on 2021/05/14.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Adrian Price-Whelan              5            33             18            0.01
Arjun Savel                     18           220             24            0.03
Christina Hedges                 5            91             82            0.02
Dan F-M                        560        165043         316130           58.16
Dan Foreman-Mackey              36          3538         176542           21.77
Eric Agol                        1             1              1            0.00
Ian Czekala                      5           115             23            0.02
Luke Bouma                       1            13              9            0.00
Rodrigo Luger                   10        164894            273           19.96
Timothy D Brandt                 1           254              0            0.03
Tom Barclay                      1             1              1            0.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Adrian Price-Whelan           1            3.0         12.4                0.00
Arjun Savel                  52           23.6          5.2                5.77
Christina Hedges             38           41.8          2.1                7.89
Dan F-M                     135            0.1          8.5               74.07
Dan Foreman-Mackey         6173          174.5         18.2               11.45
Eric Agol                     1          100.0          0.2              100.00
Ian Czekala                  70           60.9         23.8               14.29
Luke Bouma                   13          100.0         10.0                0.00
Rodrigo Luger               221            0.1         19.7                9.95
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5281/zenodo.1998447 is OK
- 10.5281/zenodo.4695331 is OK
- 10.7717/peerj-cs.55 is OK
- 10.21105/joss.01143 is OK
- 10.1093/mnras/stt1435 is OK
- 10.1093/mnrasl/slt075 is OK
- 10.3847/1538-3881/aaf22f is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK
- 10.3847/1538-3881/aae8e5 is OK
- 10.3847/1538-3881/ab4fee is OK
- 10.3847/1538-3881/aa9332 is OK
- 10.3847/2515-5172/aaaf6c is OK
- 10.1051/0004-6361/201118085 is OK
- 10.1093/mnras/stz2870 is OK
- 10.1086/669497 is OK
- 10.1093/mnras/stz2688 is OK
- 10.3847/1538-4365/abe70e is OK
- 10.1086/683602 is OK
- 10.1088/1538-3873/aaaaa8 is OK
- 10.3847/1538-3881/ab6663 is OK
- 10.1051/0004-6361/201628579 is OK
- 10.1093/mnras/sty2472 is OK
- 10.1093/mnras/stv894 is OK
- 10.1093/mnras/stz3251 is OK
- 10.3847/1538-4357/abc686 is OK
- 10.3847/1538-3881/abbc16 is OK
- 10.18637/jss.v076.i01 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.3847/1538-4357/abebe3 is OK
- 10.3847/1538-3881/aba4b2 is OK
- 10.1038/s41586-020-2400-z is OK
- 10.1093/mnras/stx138 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03285/joss.03285/10.21105.joss.03285.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03285/joss.03285/10.21105.joss.03285.pdf) :page_facing_up: :point_left:

**Comment by benjaminpope** [comment]
**Intro Comments:**
- Thanks for producing a great library, with many citations and applications already. This is already a very useful community resource. The code is well documented, with a host of great examples. I have read the paper and examples carefully, and tested several of these to the best of my poor laptop's ability. I recommend acceptance with minor revisions, providing a few comments below. 

**Installation:**
- everything installs fine with pip and with git + `python setup.py`

**Unit tests:**
- All passed in my default Python environment except for `test_small_star` and `test_sky_coords`. This was because I hadn't yet installed batman - if these are going to be in unit tests perhaps they should also be in dependencies.

**Case Studies:**
- Ran all the case studies locally on my machine by copying and pasting code. A couple worked fine, the others seemed to work ok but had very long MCMC chains and I didn't hang around to check their output. Could be neat to have these as notebooks in a notebooks/ directory on GitHub, as you just know people (read: me) are going to hack these as recipes to run on their own data.
	- Just a note - check they all import matplotlib.pyplot as plt somewhere at the top? A couple of these don't and I needed to add this line. Perhaps the GitHub actions are importing this already?
	- It could also be nice to print run times on the case studies, so that prospective users know roughly how long each stage should take. Some of the MCMC chains and optimizations take a while to compute (of course), but as users are choosing between a few alternative packages, it will be helpful to know how this performs.
	- There are many warnings: perhaps in the interests of a reader hoping to clone code there could be a warnings filter placed at the top. For example, I get a lot of (and this may be to do with my Python environment):
		- `WARNING (theano.tensor.opt): Cannot construct a scalar test value from a test value with no size: InplaceDimShuffle{x,0}.0`
		- `WARNING (theano.tensor.opt): Cannot construct a scalar test value from a test value with no size: Elemwise{cos,no_inplace}.0`
- I'm actually a little uncertain about the boundary between Tutorial and Case Study. But I note that there is an excellent astrometry Tutorial (as long as the case studies and much longer than some of the tutorials), but no analogous Case Study doing astrometric fitting.
- _Gaussian process models for stellar variability_
	- Might like to add a sentence about why VI and MCMC retrieve such different posteriors?
- _Putting it all together:_
	- Shouldn't this come last? Currently second in the list.
- _RVs with multiple instruments_
	- Mine crashed on multiprocessing sampling. Perhaps default in any notebook version should be 1 core?

**General paper comments:**
- It is not unusual, though far from universal, for astronomy code in JOSS to be illustrated with a figure or two. (Around half of the recent papers I checked at random seem to have one). Perhaps the authors might like to include one in this paper, to illustrate the diverse capabilities of `exoplanet`. The multi-instrument light curve and RV fits from "Fitting light curves from multiple instruments" and "RVs with multiple instruments" respectively could go well as a multi-panel figure.  (This is by no means a requirement for the paper to be accepted, just an opportunity). 
- The section on why autodiff is important in data analysis could perhaps do with some expansion. Not all astronomers are familiar with this, and perhaps another sentence or two selling and explaining the basic ideas of autodiff might help. (In my experience, many people do not realize it is different from finite differences). It might also do good to say explicitly whether it permits both forward and reverse mode autodiff, briefly highlight the advantages of Hamiltonian Monte Carlo, highlight the possibility of interoperability with neural network models, and so forth. I would even suggest putting a sentence in the Summary about the advantages of autodiff to hammer the point home. This is also not a required edit, but a suggestion to the authors on how to distinguish this work clearly.
- Citation formatting: fix brackets in "(based on the method used by the Stan project Carpenter et al.,2017.)" Similar problem with Theano, Theano Development Team, comma should be semicolon. Suggest semicolon before Carpenter would do?Should have citation to Jax. 
- Citations - I know everyone has their homebrew exoplanet tool kit (which is why this project is so valuable) and that the Similar Tools section is is "far from a comprehensive list", but the paper could probably afford to cite more widely in this section. Perhaps a few other tools deserve a mention: for example Parviainen's PyExoTk or the Parviainen & Aigrain `ldtk` (I list these because they are the ones I have used myself; wide citation is to be encouraged, and perhaps the authors can think of others). Given the detached EB model, perhaps a citation to Irwin's `eb` or the `PHOEBE` or `JKTEBOP` projects?
- "finite-difference gradients which can be subject to significant numerical errors and require 2N computations of a models with N free parameters" - might be good to cite and quantify this statement, while contrasting the scaling behaviour with dimensionality of proper autodiff.

**Comment by dfm** [comment]
Thanks @benjaminpope for your detailed and constructive review - all of these suggestions are very helpful!! As you've seen, I've created a series of issue threads to track these changes and I'll report back here once I've gone through them.

**Comment by grburgess** [comment]
## general

This is a very well written package from a software point of view. It follows all the modern standards of CI and maintainable python code. The documentation appears complete and testing is well done. 

I will second comments of @benjaminpope in order to not duplicate with an emphasis on expanding the autodiff explanation. Even more, why do you need HMC to do this? I wholeheartedly agree, but its a teaching opportunity to the field that these types of integrators are more robust for heavy-tailed and correlated posteriors and basically the only way to deal with high-dimensions. This seems to be a core benefit of this tool to the field, and could use a little more promotion. 

I recommend publication as well, after addressing the issues I have linked above.


**pet peeve:** In the documentation there are several references to "finding the best parameters." I know what is heuristically meant, but perhaps one could use "estimating/computing the posterior" or "conditioning the model on data"? There is a perception in the field that Bayesian inference is a way to "find the parameter errors" more robustly than MLE methods, and careful use of language can help to overcome that.. but this is a style preference only :)


### Questions

* I noticed that one of the models was an interpolated light curve. Is this actually used in fitting? If it is used as a template where the shape does not change except for the scaling, this should be fine. But HMC typically goes bonkers with interpolated models that change their shape. This is more of curiosity of mine... but I also may have not understood what that model is used for in the library.

* As a non-expert, I see this as a framework that includes some models and can be built upon. Is this a correct interpretation? Have you experimented with different parameterizations of the models and/or have suggestions for users. HMC is great, but it can still be hampered by poor parameterization. I do note that scale parameters are "logged" as they should be (in general).

**Comment by dfm** [comment]
@grburgess: Thanks for this review and the opened issues! Can I ask you to expand a bit on what you mean by:

> But HMC typically goes bonkers with interpolated models that **change their shape**.

And especially by "change their shape". I haven't found major issues with using interpolated models and I don't see any fundamental issue with using interpolation, so I'd love to hear more about your experiences!

**Comment by grburgess** [comment]
@dfm In the past in Stan, when trying to use high-D interpolation, I've found that HMC gets really confused around the nodes. But, 1D interpolation, where the coefficients are precomputed and used like a template works fine. Example, you interpolate a light curve outside of the program from some complex model, and then are simply fitting for an amplitude that scales the interpolation. But I've had issues in the past were you also try to scale the time axis (stretching). I'm sure it has something to do with HMC diverging... but I never really investigated further.

 I was not sure what the intended use was here. If it is working that is great.

**Comment by grburgess** [comment]
@arfon I am still unable to interact with the checkboxes.

**Comment by arfon** [comment]
@whedon re-invite @grburgess as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@grburgess please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by arfon** [comment]
@grburgess – could you try clicking this link and accepting the invite? https://github.com/openjournals/joss-reviews/invitations

**Comment by benjaminpope** [comment]
Re Whedon post - just waiting for confirmation on all the pull requests that changes have been made. Everything looks good so far - the comment about paper content & adding a figure (which is not compulsory at all) hasn't been decided either way, so as far as I can tell I'm just waiting on that.

**Comment by dfm** [comment]
Yes - Definitely in my court still!

Finishing all the changes will probably take me a while still, but I'll have updates on individual issues as we go. Thanks again!

**Comment by dfm** [comment]
@benjaminpope, @grburgess: Thanks again so much for your thorough reviews! I've gone through and addressed all of your comments, I think. Here's [the project showing all the related issues and PRs](https://github.com/orgs/exoplanet-dev/projects/2). The big ones were:

- https://github.com/exoplanet-dev/exoplanet/pull/186: Outlines @benjaminpope's suggestions and links to specific issues or PRs for each one.
- https://github.com/exoplanet-dev/exoplanet/pull/200: Includes all the major updates to the tutorials on the main docs page.
- https://github.com/exoplanet-dev/case-studies/pull/21: Includes all of the infrastructure updates for the new case studies.

Overall the major changes were to the documentation, so:

- I completely overhauled the [documentation page](https://docs.exoplanet.codes/en/latest/) and [case studies](https://gallery.exoplanet.codes/). The big conceptual difference is that I moved some of the more specific tutorials to the case studies page and then added some more high level tutorials about [autodiff](https://docs.exoplanet.codes/en/latest/tutorials/autodiff/), [the range of models supported by the code](https://docs.exoplanet.codes/en/latest/tutorials/data-and-models/), and [some (limited) notes about reparameterization](https://docs.exoplanet.codes/en/latest/tutorials/reparameterization/).

- I updated the infrastructure to use the `sphinx-book-theme` for both the docs and case studies which includes buttons for launching the ipynbs using binder or downloading them directly. This should make it easier for users to dive right in without having to copy and paste the code and worry about missing imports.

- I dealt with (and tried to solve) various compatibility (ugh macos multiprocessing!) and runtime issues, added some more inline comments, and fixed various small bugs or issues that were identified.

Let me know if you see any remaining issues!!

**Comment by dfm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03285/joss.03285/10.21105.joss.03285.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03285/joss.03285/10.21105.joss.03285.pdf) :page_facing_up: :point_left:

**Comment by benjaminpope** [comment]
@dfm - thanks for the updates. I have looked through the paper - happy with the new density of citations and the new figures, it's great. I have also looked at the new case studies and autodiff page - these are some of the most professional tutorial pages I have ever seen. I will be linking people to the autodiff one especially.

With that, all requested changes are complete and to a very high standard, and I am happy to recommend the paper for acceptance by JOSS.

**Comment by grburgess** [comment]
@dfm I second acceptance. The docs look fantastic and very useful. Thanks for taking the time to make this project pedagogical as well as a beneficial to the community.

**Comment by arfon** [comment]
@dfm – At this point could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by dfm** [comment]
@arfon: Thanks! I'm waiting for a final round of comments from co-authors, but I'll tag a release next week.

**Comment by dfm** [comment]
@arfon: I've bumped the version number to `v0.5.1` and the archive is on Zenodo with the correct metadata at `10.5281/zenodo.5006965`. Thanks!

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.5006965 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5006965" target="_blank">10.5281/zenodo.5006965</a> is the archive.

**Comment by arfon** [comment]
@whedon set v0.5.1 as version

**Comment by whedon** [comment]
OK. v0.5.1 is the version.

**Comment by arfon** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5281/zenodo.1998447 is OK
- 10.5281/zenodo.4695331 is OK
- 10.7717/peerj-cs.55 is OK
- 10.21105/joss.01143 is OK
- 10.1093/mnras/stt1435 is OK
- 10.1093/mnrasl/slt075 is OK
- 10.3847/1538-3881/aaf22f is OK
- 10.1051/0004-6361/201322068 is OK
- 10.3847/1538-3881/aabc4f is OK
- 10.3847/1538-3881/aae8e5 is OK
- 10.3847/1538-3881/ab4fee is OK
- 10.3847/1538-3881/aa9332 is OK
- 10.3847/2515-5172/aaaf6c is OK
- 10.1051/0004-6361/201118085 is OK
- 10.1093/mnras/stz2870 is OK
- 10.1086/669497 is OK
- 10.1093/mnras/stz2688 is OK
- 10.3847/1538-4365/abe70e is OK
- 10.1086/683602 is OK
- 10.1088/1538-3873/aaaaa8 is OK
- 10.3847/1538-3881/ab6663 is OK
- 10.1051/0004-6361/201628579 is OK
- 10.1093/mnras/sty2472 is OK
- 10.1093/mnras/stv894 is OK
- 10.1093/mnras/stz3251 is OK
- 10.3847/1538-4357/abc686 is OK
- 10.3847/1538-3881/abbc16 is OK
- 10.18637/jss.v076.i01 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.3847/1538-4357/abebe3 is OK
- 10.3847/1538-3881/aba4b2 is OK
- 10.1038/s41586-020-2400-z is OK
- 10.1093/mnras/stx138 is OK
- 10.1093/mnras/stv1857 is OK
- 10.1088/0004-637X/742/2/123 is OK
- 10.1111/j.1365-2966.2004.07871.x is OK
- 10.3847/1538-4365/abb4e2 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2400

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2400, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1407288394711961606) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2402
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03285](https://doi.org/10.21105/joss.03285)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arfon** [comment]
@grburgess, @benjaminpope – many thanks for your reviews here! JOSS relies upon volunteer efforts from people like you and we wouldn't be able to do this without you! ✨ 

@dfm – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03285/status.svg)](https://doi.org/10.21105/joss.03285)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03285">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03285/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03285/status.svg
   :target: https://doi.org/10.21105/joss.03285
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03285/status.svg)](https://doi.org/10.21105/joss.03285)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
